okio/jvm/src/main/java/okio/SelectRunner.kt

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio
+
+import java.io.IOException
+import java.nio.channels.ClosedChannelException
+import java.nio.channels.SelectableChannel
+import java.nio.channels.SelectionKey
+import java.nio.channels.Selector
+import java.util.concurrent.ConcurrentLinkedQueue
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+suspend fun <T> SelectableChannel.selectAsync(ops: Int, block: () -> T): T {
+  val selectRunner = SelectRunner.instance
+  // TODO: cancel the key when the coroutine is canceled?
```

Comment:
Did you have any ideas for how to implement this? `Task` could probably vend a `CompletionHandler` that we could directly pass to `invokeOnCancellation`.

I don't know NIO that well, but according to the docs canceling a key can potentially block, so we probably shouldn't cancel the key direction from that handler (`CompletionHandler`s are invoked synchronously). Maybe we could enqueue a special "cancelation task" that would actually cancel the key in the future? I think it's fine if the key/channel doesn't get canceled _immediately_ as long as it doesn't try to do any wasted IO in the future (which we could ensure using our own flag as discussed below).

Reply:
No cancelation task required, if we just set the flag on the existing task then wake up the selector the task can just cancel itself.